components/layout/display_list_builder.rs

```diff
@@ -1414,22 +1415,56 @@ impl FragmentDisplayListBuilding for Fragment {
     fn build_display_list_for_background_gradient(&self,
                                                   state: &mut DisplayListBuildState,
                                                   display_list_section: DisplayListSection,
-                                                  absolute_bounds: &Rect<Au>,
+                                                  absolute_bounds: Rect<Au>,
                                                   clip: &LocalClip,
                                                   gradient: &Gradient,
                                                   style: &ComputedValues,
                                                   index: usize) {
+        // Calculate where the first "tile" needs to be placed on one axis.
+        // * base is the beginning of the visible area
+        // * start is the current "tile" position
+        // * tile is the length of the "tile"
+        // Returns a difference between start and new start.
+        // It holds that base - tile < start - diff <= base
+        fn get_first_tile(base: i32, start: i32, tile: i32) -> i32 {
```

Comment:
Perhaps this should take and return `Au` instead of `i32`.  This would be more straightforward with servo/app_units#40.